### PR TITLE
feat: version compatible with vue 3

### PR DESCRIPTION
**Pull request by elevatebart**
Hello @koca ,

Here is the first draft of the prism editor working with vue 3.
I have been asked to startup a vue-live engine for vue 3.

Would you be open to delivering this one on the `@next` channel?

Do you need anything more?

### CODE REVIEW

**Comment by koca** [comment]
Hey @elevatebart! Thank you for the PR ♥️ i will review this after work(today).

**Comment by elevatebart** [comment]
I just found out that the `@testing-library/vue` was indirectly dependent on vue 2 (through their version of the vue test utils).
It means that the unit tests will not work until they deliver their upgrade. 

Could we to deliver this version as an alpha?

**Comment by koca** [comment]
Why do i see my old commits? There should be only 1 commit i think 🤔

**Comment by elevatebart** [comment]
Yep @koca I forgot to mention that.

Note that my PR is based on the `next` branch.

The `next` branch is behind `master` by those 20 commits.
I started by rebasing `master` into `next` before starting the refactor.

The only change that matters here is the last commit:

https://github.com/koca/vue-prism-editor/pull/95/commits/394b1ce6f827e9b2d262fea84049363ba9188bee

**Comment by koca** [comment]
can you rebase this branch and update the PR?

**Comment by koca** [comment]
oh i see its not from master branch

File: packages/vue-prism-editor/src/Editor.ts [hunk 1]
```diff
@@ -92,9 +93,9 @@ export const PrismEditor = Vue.extend({
     };
   },
   watch: {
-    value: {
+    modelValue: {
       immediate: true,
-      handler(newVal: string): void {
+      handler(this: any, newVal: string): void {
```
Comment (koca) [inline packages/vue-prism-editor/src/Editor.ts:98]: why did you add `this:any` to some methods?

File: packages/vue-prism-editor/src/Editor.ts [hunk 2]
```diff
@@ -104,31 +105,31 @@ export const PrismEditor = Vue.extend({
     },
     content: {
       immediate: true,
-      handler(): void {
+      handler(this: any): void {
```
Comment (koca) [inline packages/vue-prism-editor/src/Editor.ts:108]: same `this:any`

**Review by koca — CHANGES_REQUESTED** [review]

File: packages/vue-prism-editor/src/Editor.ts [hunk 3]
```diff
@@ -1,17 +1,18 @@
...
+const KEYCODE_TAB = 'Tab';
+const KEYCODE_BACKSPACE = 'Backspace';
+const KEYCODE_Y = 'y';
+const KEYCODE_Z = 'z';
+const KEYCODE_M = 'm';
+const KEYCODE_PARENS = '(';
+const KEYCODE_BRACKETS = '{';
+const KEYCODE_QUOTE = '"';
+const KEYCODE_BACK_QUOTE = '`';
+const KEYCODE_ESCAPE = 'Escape';
```
Comment (koca) [inline packages/vue-prism-editor/src/Editor.ts:15]: these are not keycodes anymore what was the problem?

Comment (elevatebart) [reply packages/vue-prism-editor/src/Editor.ts:15]: `event.keyCode` is deprecated
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

and `event.key` seems compatible with IE9+ 
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key

It has nothing to do with vue 3 and I should remove this change from this PR.

Comment (koca) [reply packages/vue-prism-editor/src/Editor.ts:15]: i think its okay we can keep the changes

Comment (elevatebart) [reply packages/vue-prism-editor/src/Editor.ts:15]: Screen Shot 2020-10-16 at 2 37 44 PM

**Comment by elevatebart** [comment]
Thank you @koca for the comments, 

`this: any` turned out to be unnecessary so I removed it.

the keyCode property is deprecated but should not prevent vue.3 from working so I removed this change too.

I hope it simplifies the review a little.

Sorry again for the confusion.

File: packages/vue-prism-editor/src/Editor.ts [hunk 4]
```diff
@@ -55,7 +56,7 @@ export const PrismEditor = Vue.extend({
       type: Boolean,
       default: false,
     },
-    value: {
+    modelValue: {
```
Comment (elevatebart) [inline packages/vue-prism-editor/src/Editor.ts:59]: value => modelValue
https://v3.vuejs.org/guide/migration/v-model.html#migration-strategy

File: packages/vue-prism-editor/src/Editor.ts [hunk 5]
```diff
@@ -261,7 +262,7 @@ export const PrismEditor = Vue.extend({
       input.selectionStart = record.selectionStart;
       input.selectionEnd = record.selectionEnd;
 
-      this.$emit('input', record.value);
+      this.$emit('update:modelValue', record.value);
```
Comment (elevatebart) [inline packages/vue-prism-editor/src/Editor.ts:265]: `this.$emit('input' => 'update:modelValue'`

https://v3.vuejs.org/guide/migration/v-model.html#migration-strategy

File: packages/vue-prism-editor/src/Editor.ts [hunk 6]
```diff
@@ -306,13 +307,11 @@ export const PrismEditor = Vue.extend({
       // console.log(navigator.platform);
       const { tabSize, insertSpaces, ignoreTabKey } = this;
 
-      if (this.$listeners.keydown) {
-        // onKeyDown(e);
-        this.$emit('keydown', e);
+      // onKeyDown(e);
+      this.$emit('keydown', e);
```
Comment (elevatebart) [inline packages/vue-prism-editor/src/Editor.ts:311]: `this.$listeners` has been removed in vue 3 (or is null as far as I can tell) which triggers an undefined error.

File: packages/vue-prism-editor/src/Editor.ts [hunk 7]
```diff
@@ -512,27 +511,23 @@ export const PrismEditor = Vue.extend({
       }
     },
   },
-  render(h): VNode {
+  render() {
```
Comment (elevatebart) [inline packages/vue-prism-editor/src/Editor.ts:514]: h is now imported from vue

File: packages/vue-prism-editor/src/Editor.ts [hunk 8]
```diff
@@ -512,27 +511,23 @@ export const PrismEditor = Vue.extend({
...
+        class: 'prism-editor__line-width-calc',
+        style: 'height: 0px; visibility: hidden; pointer-events: none;',
       },
       '999'
     );
     const lineNumbers = h(
       'div',
       {
-        staticClass: 'prism-editor__line-numbers',
+        class: 'prism-editor__line-numbers',
```
Comment (elevatebart) [inline packages/vue-prism-editor/src/Editor.ts:526]: h has changed

https://v3.vuejs.org/guide/migration/render-function-api.html#vnode-props-format

File: packages/vue-prism-editor/src/Editor.ts (hunk 6 above)
Comment (elevatebart) [reply packages/vue-prism-editor/src/Editor.ts:311]: After investigation, `this.$listeners` exists only if it has been implemented, with an `@click=` or an `@keydown=` 
It should be protected against null.
Since it is deprecated anyway, I removed it completely.

**Comment by elevatebart** [comment]
I have added comments in this commit to explain my changes

https://github.com/koca/vue-prism-editor/pull/95/commits/77bb9af05a21e1536add2233192fef0e67c50e9c

File: packages/vue-prism-editor/src/Editor.ts (hunk 6 above)
Comment (koca) [reply packages/vue-prism-editor/src/Editor.ts:311]: yep they've merged `$listeners` to `$attrs`. So `$attrs` has both `$attrs` and `$listeners` in vue 3

**Comment by koca** [comment]
There is only console.log left can you please remove it and i've tested on my local, seems to be working fine. We can't run tests but I'll publish the alpha version for you. Thank you @elevatebart 😊👍

**Review by koca — APPROVED** [review]

**Comment by koca** [comment]
Hey @elevatebart just released `v2.0.0-alpha.1`

example vue 3 csb: https://codesandbox.io/s/vue-3-support-for-vue-prism-editor-fm7v5

install:

```sh
npm install vue-prism-editor@alpha
```

or

```sh
yarn add vue-prism-editor@alpha
```

cheers 🎉
